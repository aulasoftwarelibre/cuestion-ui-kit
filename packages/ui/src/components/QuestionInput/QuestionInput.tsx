import {
  AppBar,
  Avatar,
  Button,
  Grid,
  Paper,
  TextField,
  Toolbar,
} from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";

import * as React from "react";
import { FormattedMessage, InjectedIntl, injectIntl } from "react-intl";

import { messages } from "@cuestion/common";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    sendButton: {
      marginRight: theme.spacing(1),
    },
    block: {
      display: "block",
    },
    contentWrapper: {
      margin: "16px 16px",
    },
    paper: {
      margin: "auto",
      maxWidth: 936,
      overflow: "hidden",
    },
    searchBar: {
      borderBottom: "1px solid rgba(0, 0, 0, 0.12)",
    },
    searchInput: {
      fontSize: theme.typography.fontSize,
    },
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: "100%",
    },
  }),
);

interface Props {
  handleOnSubmit: any;
  intl: InjectedIntl;
  length: number;
}

interface Question {
  user: string;
  question: string;
}

const _QuestionInput: React.FunctionComponent<Props> = ({
  handleOnSubmit,
  intl,
  length,
}) => {
  const [enabled, setEnabled] = React.useState<boolean>(false);
  const [data, setData] = React.useState<Question>({
    user: "",
    question: "",
  });
  const classes = useStyles();

  const onSubmit = () => {
    if (data.question.length < length) {
      return;
    }

    handleOnSubmit({
      ...data,
    });

    setEnabled(false);

    setData({
      user: "",
      question: "",
    });
  };

  const handleOnChange = () => event => {
    setData({
      ...data,
      [event.target.name]: event.target.value,
    });

    if (event.target.name === "question") {
      setEnabled(event.target.value.length >= length);
    }
  };

  return (
    <Paper className={classes.paper}>
      <div className={classes.contentWrapper}>
        <TextField
          id="question"
          className={classes.textField}
          helperText={intl.formatMessage(messages.enterLengthCharacters, {
            length,
          })}
          label={intl.formatMessage(messages.askQuestion, {})}
          name="question"
          margin="normal"
          multiline
          onChange={handleOnChange()}
          rows={5}
          value={data.question}
        />
      </div>
      <AppBar
        className={classes.searchBar}
        position="static"
        color="default"
        elevation={0}>
        <Toolbar>
          <Grid container spacing={2} alignItems="center">
            <Grid item>
              <Avatar>{data.user ? data.user[0] : ""}</Avatar>
            </Grid>
            <Grid item xs>
              <TextField
                id="user"
                fullWidth
                placeholder={intl.formatMessage(messages.yourName, {})}
                InputProps={{
                  className: classes.searchInput,
                  disableUnderline: true,
                }}
                name="user"
                value={data.user}
                onChange={handleOnChange()}
              />
            </Grid>
            <Grid item>
              <Button
                id="send"
                variant="contained"
                color="primary"
                onClick={() => onSubmit()}
                disabled={!enabled}>
                <FormattedMessage {...messages.send} />
              </Button>
            </Grid>
          </Grid>
        </Toolbar>
      </AppBar>
    </Paper>
  );
};

export const QuestionInput = injectIntl(_QuestionInput);

export default QuestionInput;
