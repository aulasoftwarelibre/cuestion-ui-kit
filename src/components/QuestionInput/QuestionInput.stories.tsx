import { action } from "@storybook/addon-actions";
import { withKnobs } from "@storybook/addon-knobs";
import { storiesOf } from "@storybook/react";
import * as React from "react";

import { wInfo } from "../../utils/wInfo";
import QuestionInput from "./QuestionInput";

const stories = storiesOf("Components/QuestionInput", module);
stories.addDecorator(withKnobs);

stories.add("default", wInfo()(() => <QuestionInput handleOnSubmit={action("question")} />));
