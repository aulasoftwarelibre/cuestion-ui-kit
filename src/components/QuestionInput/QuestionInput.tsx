import { AppBar, Avatar, Button, Grid, Paper, TextField, Toolbar } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";

import * as React from "react";
import { FormattedMessage, InjectedIntl, injectIntl } from "react-intl";

import messages from "../../languages/messages";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    sendButton: {
      marginRight: theme.spacing(1)
    },
    block: {
      display: "block"
    },
    contentWrapper: {
      margin: "16px 16px"
    },
    paper: {
      margin: "auto",
      maxWidth: 936,
      overflow: "hidden"
    },
    searchBar: {
      borderBottom: "1px solid rgba(0, 0, 0, 0.12)"
    },
    searchInput: {
      fontSize: theme.typography.fontSize
    },
    textField: {
      marginLeft: theme.spacing(1),
      marginRight: theme.spacing(1),
      width: "100%"
    }
  })
);

interface Props {
  handleOnSubmit: any;
  intl: InjectedIntl;
}

interface Question {
  user: string;
  question: string;
}

const _QuestionInput: React.FunctionComponent<Props> = ({ handleOnSubmit, intl }) => {
  const [enabled, setEnabled] = React.useState<boolean>(false);
  const [data, setData] = React.useState<Question>({
    user: "",
    question: ""
  });
  const classes = useStyles();

  const onSubmit = () => {
    if (data.question.length < 10) {
      return;
    }

    handleOnSubmit({
      ...data
    });

    setEnabled(false);

    setData({
      user: "",
      question: ""
    });
  };

  const handleOnChange = () => event => {
    setData({
      ...data,
      [event.target.name]: event.target.value
    });

    if (event.target.name === "question") {
      setEnabled(event.target.value.length >= 10);
    }
  };

  return (
    <Paper className={classes.paper}>
      <div className={classes.contentWrapper}>
        <TextField
          id="question"
          className={classes.textField}
          label={intl.formatMessage(messages.askQuestion, {})}
          name="question"
          multiline
          rows={5}
          value={data.question}
          onChange={handleOnChange()}
          margin="normal"
        />
      </div>
      <AppBar className={classes.searchBar} position="static" color="default" elevation={0}>
        <Toolbar>
          <Grid container spacing={2} alignItems="center">
            <Grid item>
              <Avatar>{data.user ? data.user[0] : ""}</Avatar>
            </Grid>
            <Grid item xs>
              <TextField
                id="user"
                fullWidth
                placeholder={intl.formatMessage(messages.yourName, {})}
                InputProps={{
                  className: classes.searchInput,
                  disableUnderline: true
                }}
                name="user"
                value={data.user}
                onChange={handleOnChange()}
              />
            </Grid>
            <Grid item>
              <Button variant="contained" color="primary" onClick={() => onSubmit()} disabled={!enabled}>
                <FormattedMessage {...messages.send} />
              </Button>
            </Grid>
          </Grid>
        </Toolbar>
      </AppBar>
    </Paper>
  );
};

export const QuestionInput = injectIntl(_QuestionInput);

export default QuestionInput;
