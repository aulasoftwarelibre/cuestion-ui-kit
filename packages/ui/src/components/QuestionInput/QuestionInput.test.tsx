import { Button, TextField } from "@material-ui/core";

import * as React from "react";

import {
  createComponentWithIntl,
  mountWithIntl,
} from "../../utils/createComponentWithIntl";
import QuestionInput from "./QuestionInput";

describe("QuestionInput", () => {
  const mockOnHandleSubmit = jest.fn();

  it("should render correctly", () => {
    const component = createComponentWithIntl(
      <QuestionInput handleOnSubmit={mockOnHandleSubmit} length={10} />,
    ).toJSON();

    expect(component).toMatchSnapshot();
  });

  it("should tell to press SEND", () => {
    const wrapper = mountWithIntl(
      <QuestionInput handleOnSubmit={mockOnHandleSubmit} length={10} />,
    );
    const input = wrapper.find(TextField).find("input");

    input.simulate("change", {
      target: { name: "question", value: "abcdefghij" },
    });

    expect(wrapper.find(Button).prop("disabled")).toBeFalsy();
  });

  it("should wait to tell press SEND", () => {
    const wrapper = mountWithIntl(
      <QuestionInput handleOnSubmit={mockOnHandleSubmit} length={12} />,
    );
    const input = wrapper.find(TextField).find("input");

    input.simulate("change", {
      target: { name: "question", value: "abcdefghij" },
    });

    expect(wrapper.find(Button).prop("disabled")).toBeTruthy();
  });

  it("should return a Question without an user defined", () => {
    const wrapper = mountWithIntl(
      <QuestionInput handleOnSubmit={mockOnHandleSubmit} length={10} />,
    );
    const input = wrapper.find(TextField).find("input");
    const button = wrapper.find(Button);

    input.simulate("change", {
      target: { name: "question", value: "abcdefghij" },
    });

    button.simulate("click");

    expect(mockOnHandleSubmit).toBeCalledWith({
      question: "abcdefghij",
      user: "",
    });
  });

  it("should return a Question with an user defined", () => {
    const wrapper = mountWithIntl(
      <QuestionInput handleOnSubmit={mockOnHandleSubmit} length={10} />,
    );
    const input = wrapper.find(TextField).find("input");
    const button = wrapper.find(Button);

    input.simulate("change", {
      target: { name: "question", value: "abcdefghij" },
    });

    input.simulate("change", {
      target: { name: "user", value: "user123" },
    });

    button.simulate("click");

    expect(mockOnHandleSubmit).toBeCalledWith({
      question: "abcdefghij",
      user: "user123",
    });
  });
});
