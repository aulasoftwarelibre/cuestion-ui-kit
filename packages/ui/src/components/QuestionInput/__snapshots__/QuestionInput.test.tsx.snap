// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionInput should render correctly 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-4 MuiPaper-rounded"
>
  <div
    className="makeStyles-contentWrapper-3"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-textField-7 MuiFormControl-marginNormal"
    >
      <label
        className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
        data-shrink={false}
        htmlFor="question"
      >
        Ask a question
      </label>
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-multiline MuiInput-multiline"
        onClick={[Function]}
      >
        <textarea
          aria-describedby="question-helper-text"
          aria-invalid={false}
          className="MuiInputBase-input MuiInput-input MuiInputBase-inputMultiline MuiInput-inputMultiline"
          disabled={false}
          id="question"
          name="question"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          rows={5}
          value=""
        />
      </div>
      <p
        className="MuiFormHelperText-root"
        id="question-helper-text"
      >
        Enter 10 characters
      </p>
    </div>
  </div>
  <header
    className="MuiPaper-root MuiPaper-elevation0 MuiAppBar-root MuiAppBar-positionStatic makeStyles-searchBar-5 MuiAppBar-colorDefault"
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-align-items-xs-center"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <div
            className="MuiAvatar-root MuiAvatar-colorDefault"
          >
            
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true"
        >
          <div
            className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              className="MuiInputBase-root MuiInput-root makeStyles-searchInput-6 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                id="user"
                name="user"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Your name (optional)"
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-disabled Mui-disabled"
            disabled={true}
            id="send"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span>
                Send
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
